File: clojure/perf_memory/src/clj/org/perf/test2_refl.clj
;;
;; Running with Clojure 1.0 under Windows XP
;; As of 7/15/2009
;; baseline_hello.clj
;; 
;; Some examples from gnuvince:
;; http://gnuvince.wordpress.com/2009/05/11/clojure-performance-tips/
;; And Sergio
;; http://devlog.bigmonachus.org/2009/03/performance-tips-for-clojure.html
;;
;; Test Output (as of 7/15/2009):
;; time ./test2.bat
;; -----------------

;; -----------------

(ns org.perf.test2_refl
	(:use org.perf.octane_utils_common)
	(:import (java.util Date)
             (java.io File)))

(defn str-split-refl [s regex]
      (vec (. s split regex)))

(defn str-split-fast [#^String s regex]
      (vec (. s split regex)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn my-str-join
  "Returns a string of all elements in 'sequence', separated by
  'separator'.  Like Perl's 'join'. From clojure_contrib_str: "
  [separator sequence]
  ;;;;;;;;;;;;;;;;;;;;;;
  (apply str (interpose separator sequence)))

(defn build-big-str
  []
  ;;;
  (let [str-size (* 1 1024)
        n        1024
        buf-1 (StringBuffer. (* str-size n))
        buf-2 (StringBuffer.)]    
  (dotimes [_ n]
      (let [b (random-string str-size)]
        (.append buf-1 (str b "|"))))
  (.toString buf-1)))
        
(defn run-test
  " Application Entry Point"
  []
  ;;;;
  (println "Running Test [test string split and other str utils] " *current-date*)
  (println (*memory-usage*))
  (let [data (build-big-str)]
    (println "Number of chars in str : " (count data)))
  (println (*memory-usage*))
  (println "Done"))

(defn run-cool-down
  "Allow a little time to see if JVM garbage collects.
 This may not have any effect"
  []
  ;;;
  (println (*memory-usage*))
  (Thread/sleep 100)
  (println (*memory-usage*))
  (Thread/sleep 80))
 
(defn main [& args]
  (try (run-test)
       (run-cool-down)
	   (catch Exception e
			  (println "ERR at <Main [1]>: " e)
              (.printStackTrace e)
			  (exit))))

(main)                        
;; End of Script

File: trunk/clojure/perf_memory/src/java/src/org/node/perf/test/Test2Refl.java
package org.node.perf.test;

import org.node.perf.test.util.DoTimes;
import org.node.perf.test.util.SimpleDoTimes;
import org.node.perf.test.util.SimpleTime;
import org.node.perf.test.util.Test;
import org.node.perf.test.util.Times;
import org.node.perf.util.StringUtils;
import org.node.perf.util.SystemUtils;

/**
 * 
-----------------------------
Current clojure baseline test:


-----------------------------
 *
 */
public class Test2Refl implements Test {

    public static final void main(String [] args) {
        final Test test = new Test2Refl();
        test.main(args, null);
    }
        
    public void main(String [] args, Object input) {
        
        try {            
            this.runTest();
            this.runCoolDown();
            System.out.println("Done");
        } catch(Exception e) {
            e.printStackTrace();
        }
        
    }

    /////////////////////////////////////////////////////////////////
    
    public static final String buildBigStr() {
        /*
         * In Clojure, this code might look like:
         ----------------------------------------- 
        (defn build-big-str
          []
          ;;;
          (let [str-size (* 1 1024)
                n        1024
                buf-1 (StringBuffer. (* str-size n))
                buf-2 (StringBuffer.)]    
          (dotimes [_ n]
              (let [b (random-string str-size)]
                (.append buf-1 (str b "|"))))
          (.toString buf-1)))
         */
        
        int strSize = 1 * 1024;
        int n = 1024;
        final StringBuffer buf1 = new StringBuffer(n * strSize);
        final StringBuffer buf2 = new StringBuffer();
        String b = "";
        for (int i = 0; i < n; i++) {
            b = StringUtils.randomString(strSize);
            buf1.append(b).append("|");
        }
        return buf1.toString();
    }
    
/*
 
 Mimic the following clojure functionality using the Times and DoTimes interfaces:
   
    (time (dotimes [_ 10] (let [data (build-big-str)]
            (println "Number of chars in str : " (count data)))))
     
 */
    public final class Test2Times implements Times {

        public final void execute() {            
            String data = buildBigStr();
            System.out.println("Count: " + data.length());
        }
        
    } // End of Class //
    
    
    /////////////////////////////////////////////////////////////////
    
    public void runTest() {
                
        System.out.println("Running Test [test string split and other str utils] " + SystemUtils.currentDate());
        System.out.println(SystemUtils.memoryUsage());
        
        // In clojure: (time (dotimes [_ 1e6] (* 2.2 4.1 5.4)))
                
        new SimpleTime(new SimpleDoTimes(new Test2Times(), (int)10)).execute();              
        
        System.out.println(SystemUtils.memoryUsage());       
    }

 /*
  ---------------------------
  Mimic Clojure Code:
  ---------------------------
(defn run-cool-down
  "Allow a little time to see if JVM garbage collects.
 This may not have any effect"
  []
  ;;;
  (println (*memory-usage*))
  (Thread/sleep 100)
  (println (*memory-usage*))
  (Thread/sleep 80))
    
  */
    public void runCoolDown() {
        try {
            System.out.println(SystemUtils.memoryUsage());
            Thread.sleep(100);
            System.out.println(SystemUtils.memoryUsage());
            Thread.sleep(80);
        } catch(InterruptedException e) {
            
        }
    }
    
    public void runTest2() {
        
    }

    public void runTest3() {
        
    }

    public void runTest4() {        
        
    }

} // End of Class //
